Core/ThreadEventQueue.h

```diff
@@ -48,34 +50,68 @@ struct ThreadEventQueue : public B {
 	}
 
 	bool HasEvents() {
-		lock_guard guard(eventsLock_);
-		return !events_.empty();
+		if (threadEnabled_) {
+			lock_guard guard(eventsLock_);
+			return !events_.empty();
+		} else {
+			return !events_.empty();
+		}
 	}
 
 	void NotifyDrain() {
-		lock_guard guard(eventsLock_);
-		eventsDrain_.notify_one();
+		if (threadEnabled_) {
+			lock_guard guard(eventsLock_);
+			eventsDrain_.notify_one();
+		}
 	}
 
 	Event GetNextEvent() {
-		lock_guard guard(eventsLock_);
-		if (events_.empty()) {
-			NotifyDrain();
-			return EVENT_INVALID;
-		}
+		if (threadEnabled_) {
+			lock_guard guard(eventsLock_);
+			if (events_.empty()) {
+				NotifyDrain();
+				return EVENT_INVALID;
+			}
 
-		Event ev = events_.front();
-		events_.pop_front();
-		return ev;
+			Event ev = events_.front();
+			events_.pop_front();
+			return ev;
+		} else {
+			if (events_.empty()) {
+				return EVENT_INVALID;
+			}
+			Event ev = events_.front();
+			events_.pop_front();
+			return ev;
+		}
 	}
 
 	void RunEventsUntil(u64 globalticks) {
+		if (!threadEnabled_) {
+			do {
+				for (Event ev = GetNextEvent(); EventType(ev) != EVENT_INVALID; ev = GetNextEvent()) {
+					switch (EventType(ev)) {
```

Comment:
I think we should move this switch to a separate method just to keep duplication down.

-[Unknown]